src/python/grpcio/tests/qps/client_runner.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2016, Google Inc.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+#     * Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above
+# copyright notice, this list of conditions and the following disclaimer
+# in the documentation and/or other materials provided with the
+# distribution.
+#     * Neither the name of Google Inc. nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import abc
+import thread
+import time
+
+
+class ClientRunner:
+  """Abstract interface for sending requests from clients."""
+
+  __metaclass__ = abc.ABCMeta
+
+  def __init__(self, client):
+    self._client = client
+
+  @abc.abstractmethod
+  def start(self):
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def stop(self):
+    raise NotImplementedError()
+
+
+class OpenLoopClientRunner(ClientRunner):
+
+  def __init__(self, client, interval_generator):
+    super(OpenLoopClientRunner, self).__init__(client)
+    self._is_running = False
+    self._interval_generator = interval_generator
+
+  def start(self):
+    self._is_running = True
+    self._client.start()
+    thread.start_new_thread(self._dispatch_requests, ())
+
+  def stop(self):
+    self._is_running = False
+    self._client.stop()
+    self._client = None
+
+  def _dispatch_requests(self):
+    while self._is_running:
+      self._client.send_request()
+      time.sleep(next(self._interval_generator))
```

Comment:
I'd recommend adding a docstring on BenchmarkClient's abstract send_request method rather than at module level, especially if you foresee no new client types needing to employ blocking behavior.